src/librustc_const_eval/eval.rs

```diff
@@ -569,12 +524,21 @@ fn resolve_trait_associated_const<'a, 'tcx: 'a>(
         // when constructing the inference context above.
         match selection {
             traits::VtableImpl(ref impl_data) => {
-                let name = tcx.associated_item(trait_item_id).name;
+                let name = trait_item.name;
                 let ac = tcx.associated_items(impl_data.impl_def_id)
                     .find(|item| item.kind == ty::AssociatedKind::Const && item.name == name);
                 match ac {
-                    Some(ic) => lookup_const_by_id(tcx, ic.def_id, Substs::empty()),
-                    None => default_value,
+                    Some(ic) => {
+                        let substs = tcx.lift(&impl_data.substs).unwrap();
```

Comment:
There used to be a `drain_fulfillment_cx` without the panic part function. Can you reintroduce it? We need to not drop nested obligations on the floor, because of associated types in impls:

```Rust
trait Foo {
    const BAR: usize;
}

impl<U, V> Foo for (U, V) where U: Iterator<Item=V>, V: Foo {
    const BAR: usize = V::BAR;
}

impl Foo for u32 { const BAR: usize = 4; }
```